### PR TITLE
Issue 14: Add function to generate point nowcast from delay

### LINKED ISSUE #14: Write function to generate point nowcast
**Goal**
`get_delay_estimate()` is an exported package function that returns a `delay_df` which contains an empirical estimate of the delay pmf (proportion of all cases that were reported with delay $d$) alongside the corresponding delay $d$. The next step is to use this on a reporting triangle to generate a point nowcast. This will then be performed repeatedly to estimate the observation error. It could be an exported function, or we could keep it internal and only export the function that generates the complete probabilistic nowcast. 

**Solution**
Write an `apply_delay()` function which iteratively updates the reporting triangle to generate a reporting square (bottom right of the triangle repeats) using the delay pmf. 

**Describe alternatives you've considered**
An alternative approach would have been to generate the point nowcast in the same step as estimating the delay. We have intentionally chosen to separate these two.

### PR DESCRIPTION
**Pull request by kaitejohnson**
## Description
This PR closes #14 and #23

This PR writes a new function to apply a delay pmf to a reporting triangle to "fill in" the reporting triangle. It also uses this function in the Getting Started vignette (though we may eventually want to return this to an internal function to encourage users to generate probabilistic nowcasts).  It adds some tests to this function, though we will still want to add more once we add more validations. 

In the vignette, it then compares the point nowcast to the later observed data points and the data without nowcast (just visually)

Scope creep:
 - adds tests to `handle_neg_vals()` function

Note, am trying to intentionally keep these PRs really small (so one function/new functionality at a time). Then once everything "works" for one example we can define more concretely the additional functionalities/edge cases we want to be able to handle. 

Here's the example of the point nowcast from the Rmd in blue, compared to the 3 months later observed data (black) and the data without a nowcast (red)

image

## Checklist

- [X] My PR is based on a package issue and I have explicitly linked it.
- [X] I have included the target issue or issues in the PR title in the for Issue(s) *issue-numbers*: PR title
- [X] I have read the [contribution guidelines](https://github.com/epinowcast/.github/blob/main/CONTRIBUTING.md).
- [X] I have tested my changes locally.
- [X] I have added or updated unit tests where necessary.
- [X] I have updated the documentation if required.
- [X] My code follows the established coding standards.
- [X] I have added a news item linked to this PR.
- [X] I have reviewed CI checks for this PR and addressed them as far as I am able.

### CODE REVIEW

File: R/apply_delay.R [hunk 1]
```diff
@@ -0,0 +1,65 @@
...
+  validate_delay_and_triangle(
+    triangle_to_nowcast,
+    delay_pmf
+  )
+  n_delays <- length(delay_pmf)
+  n_dates <- nrow(triangle_to_nowcast)
+  expectation <- triangle_to_nowcast
+
+  for (co in 2:n_delays) {
+    block_bottom_left <- expectation[(n_dates - co + 2):n_dates, 1:(co - 1),
```
Comment (seabbs) [inline R/apply_delay.R:53]: this internal code would be much better as a internal function that is then called. Ideally using a apply or other iterative operator?

File: vignettes/baselinenowcast.Rmd [hunk 2]
```diff
@@ -114,24 +106,29 @@ at each reference date, observed as of the "fully observed" data on October
...
+    aes(x = reference_date, y = confirm), color = "darkred"
   ) +
   geom_line(
     data = epinowcast::germany_covid19_hosp |>
-      dplyr::filter(location == "DE", age_group == "00+") |>
-      group_by(reference_date) |>
-      summarise(total_confirmed = sum(confirm)) |>
-      dplyr::filter(reference_date <= "2021-07-01"),
-    aes(x = reference_date, y = total_confirmed), color = "black"
+      dplyr::filter(
```
Comment (seabbs) [inline vignettes/baselinenowcast.Rmd:119]: I dont think we need to use package::function here its certainly not epinowcast org standards

File: vignettes/baselinenowcast.Rmd [hunk 3]
```diff
@@ -274,11 +282,59 @@ The reporting triangle we are applying it to must have the same `max_delay`
...
-# )#nolint
+point_reporting_square <- apply_delay(
+  triangle_to_nowcast = triangle,
+  delay_pmf = delay_df$pmf
+)
+```
+
+We'll make a quick plot to compare the nowcast to the later observed data
+```{r}
+# Get the confirmations through July 1, 2021, from the observations up until
```
Comment (seabbs) [inline vignettes/baselinenowcast.Rmd:293]: why comment when it could just be text

File: R/apply_delay.R (hunk 1 above)
Comment (kaitejohnson) [reply R/apply_delay.R]: I love for loops for readability but happy to follow standard here, see my edit

File: vignettes/baselinenowcast.Rmd [hunk 4]
```diff
@@ -82,25 +82,17 @@ For this quick start, we use daily level data from the
...
+
+# Filtering and plotting the data
 
 We will filter the data to just look at the national-level data, for all age
 groups.
 We will pretend that we are making a nowcast as of July 1, 2021, therefore we
 will exclude all reference dates and report dates after that date.
 `germany_covid19_hosp` is provided as package data from `epinowcast`
-```{r load-data}
-observed_long <- epinowcast::germany_covid19_hosp |>
```
Comment (seabbs) [inline vignettes/baselinenowcast.Rmd:93]: as above I dont think you need to use package function here for the epinowcast data as you already say where it is from. I would actually rename it on first talking about it to something generic to make it easier for people to copy and paste this code and use it on their own data.

File: vignettes/baselinenowcast.Rmd [hunk 5]
```diff
@@ -274,12 +282,60 @@ The reporting triangle we are applying it to must have the same `max_delay`
...
+  enw_filter_reference_dates(
+    latest_date = "2021-07-01",
+    include_days = n_history - 1
+  ) |>
+  group_by(reference_date) |>
+  summarise(
+    total_confirmed = max(confirm)
+  ) |>
+  dplyr::mutate(nowcast = rowSums(point_reporting_square))
+
```
Comment (seabbs) [inline vignettes/baselinenowcast.Rmd:306]: random extra line break

File: vignettes/baselinenowcast.Rmd [hunk 6]
```diff
@@ -274,12 +282,60 @@ The reporting triangle we are applying it to must have the same `max_delay`
...
+  enw_filter_report_dates(latest_date = "2021-10-01") |>
+  enw_filter_reference_dates(
+    latest_date = "2021-07-01",
+    include_days = n_history - 1
+  ) |>
+  group_by(reference_date) |>
+  summarise(
+    total_confirmed = max(confirm)
+  ) |>
+  dplyr::mutate(nowcast = rowSums(point_reporting_square))
```
Comment (seabbs) [inline vignettes/baselinenowcast.Rmd:305]: why is mutate package::function and nothing else is?

File: vignettes/baselinenowcast.Rmd [hunk 7]
```diff
@@ -274,12 +282,60 @@ The reporting triangle we are applying it to must have the same `max_delay`
...
+```{r}
+final_data <- epinowcast::germany_covid19_hosp[location == "DE"][age_group == "00+"] |> # nolint
+  enw_filter_report_dates(latest_date = "2021-10-01") |>
+  enw_filter_reference_dates(
+    latest_date = "2021-07-01",
+    include_days = n_history - 1
+  ) |>
+  group_by(reference_date) |>
+  summarise(
+    total_confirmed = max(confirm)
```
Comment (seabbs) [inline vignettes/baselinenowcast.Rmd:303]: if you wanted you could pull this out of the preprocess object vs recalc (i.e have a at the time and final preprocess object)

File: vignettes/baselinenowcast.Rmd [hunk 8]
```diff
@@ -274,12 +282,60 @@ The reporting triangle we are applying it to must have the same `max_delay`
...
+  ylab("Confirmed admissions") +
+  scale_y_continuous(trans = "log10") +
+  ggtitle("Comparing real-time, nowcasted, and later observed cases")
+```
+Here we can see that our point nowcast slightly overestimates what was
+eventually reported (black line), but does a decent overall job of correcting
+for the right-truncation observed in the red line (the data prior to the
+nowcast).
+
+What we of course really want is a probabilistic estimate of the nowcasted
```
Comment (seabbs) [inline vignettes/baselinenowcast.Rmd:335]: why of cours? I am new to this whole nowcasting thing and I just want my dashboard to look cool

**Review by seabbs — APPROVED** [review]
This looks good as a complete PR delivering on its mission. There are a few comments on the consistency of style that I think could be addressed here or in a separate issue. Content wise everything looks sensible. 

Just noting that it really helps review if you resolve out comments with ideally something like I have done this and I did x or why you haven't done it

File: R/apply_delay.R [hunk 9]
```diff
@@ -0,0 +1,81 @@
...
+#' library(epinowcast)
+#' nat_germany_hosp <-
+#'   germany_covid19_hosp[location == "DE"][age_group == "00+"]
+#' nat_germany_hosp <- enw_filter_report_dates(
+#'   nat_germany_hosp,
+#'   latest_date = "2021-10-01"
+#' )
+#' pobs <- enw_preprocess_data(nat_germany_hosp, max_delay = 21)
+#' triangle_raw <- pobs$reporting_triangle[[1]] |>
+#'   dplyr::select(-`.group`, -reference_date) |>
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/apply_delay.R:32]: more broadly (and related to the discussion in #12) this feels like quite a lot of overhead for a function example. We'll probably want to show the full workflow somewhere but here I think it would make sense to start from a reporting triangle that's included in the package as an example, or else one that's created by hand.

File: R/apply_delay.R [hunk 10]
```diff
@@ -0,0 +1,81 @@
...
+#' @param co An integer indicating the column index
+#' @param expectation A matrix of the partially complete reporting triangle
+#' @param n_dates An integer indicating the number of dates in the reporting
+#' triangle (number of rows in the reporting triangle)
+#' @param delay_pmf A vector specifying the probability of a case being
+#' reported with delay d
+#'
+#' @returns a matrix with another set of entries corresponding to the updated
+#' values for the specified rows and column
+calc_expectation <- function(co, expectation, n_dates, delay_pmf) {
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/apply_delay.R:76]: Also needs `@keywords internal` if not to appear in the manual

File: R/get_delay_estimate.R [hunk 11]
```diff
@@ -10,7 +10,6 @@
 #' Modified from: https://github.com/KITmetricslab/RESPINOW-Hub/blob/7cce3ae2728116e8c8cc0e4ab29074462c24650e/code/baseline/functions.R#L55 #nolint
 #' @param triangle matrix of the reporting triangle, with rows representing
 #' the time points of reference and columns representing the delays
-#' indexed at 0
 #' @param max_delay integer indicating the maximum delay to estimate, in units
 #' of the delay. The default is to use the whole reporting triangle,
 #'  `ncol(triangle) -1`.
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/get_delay_estimate.R:15]: whilst we're on formatting another point is that it's good practice to indent subsequent lines for readability (depending on which editor you use this could happen automatically?) see e.g. https://github.com/r-lib/roxygen2/blob/9652d15221109917d46768e836eaf55e33c21633/R/block.R#L14

File: R/validate.R [hunk 12]
```diff
@@ -47,3 +54,34 @@ validate_triangle <- function(triangle,
     )
   }
 }
+
+#' Validate triangle to nowcast and delay pmf together
+#' @description
+#' Various checks to make sure that the reporting triangle  and the delay pmf
+#' passed in to `apply_delay()` are formatted properly and compaitble
```
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/validate.R:61]: This is a useful reference: https://roxygen2.r-lib.org/articles/index-crossref.html

File: R/apply_delay.R [hunk 13]
```diff
@@ -0,0 +1,81 @@
+#' Apply the delay to generate a point nowcast
+#' @description
+#' This function takes as an input the reporting triangle that we want to
+#' complete with a nowcast and a delay pmf and generates a point estimate of a
+#' completed reporting square (or rectangle). This code is based on the code
```
Comment (sbfnk) [inline R/apply_delay.R:5]:
```suggestion
#' Generate a point estimate of a completed reporting square (or rectangle)
#' from a reporting triangle that we want to complete with a nowcast and a 
#' delay pmf. This code is based on the code
```
(always good to start with what the function does)

File: R/apply_delay.R [hunk 14]
```diff
@@ -0,0 +1,81 @@
...
+#' @param expectation A matrix of the partially complete reporting triangle
+#' @param n_dates An integer indicating the number of dates in the reporting
+#' triangle (number of rows in the reporting triangle)
+#' @param delay_pmf A vector specifying the probability of a case being
+#' reported with delay d
+#'
+#' @returns a matrix with another set of entries corresponding to the updated
+#' values for the specified rows and column
+calc_expectation <- function(co, expectation, n_dates, delay_pmf) {
+  block_bottom_left <- expectation[(n_dates - co + 2):n_dates, 1:(co - 1), drop = FALSE] # nolint
```
Comment (sbfnk) [inline R/apply_delay.R:77]:
```suggestion
  block_bottom_left <- expectation[
    (n_dates - co + 2):n_dates,
    1:(co - 1), 
    drop = FALSE
  ]
```
maybe just me but I would find this easier to read

File: R/validate.R [hunk 15]
```diff
@@ -47,3 +54,34 @@ validate_triangle <- function(triangle,
     )
   }
 }
+
+#' Validate triangle to nowcast and delay pmf together
+#' @description
+#' Various checks to make sure that the reporting triangle  and the delay pmf
```
Comment (sbfnk) [inline R/validate.R:60]:
```suggestion
#' Various checks to make sure that the reporting triangle and the delay pmf
```

File: R/apply_delay.R [hunk 16]
```diff
@@ -0,0 +1,81 @@
...
+#' @description
+#' This function takes as an input the reporting triangle that we want to
+#' complete with a nowcast and a delay pmf and generates a point estimate of a
+#' completed reporting square (or rectangle). This code is based on the code
+#' originally developed by the Karlsruhe Institute of Technology RESPINOW
+#' German Hospitalization Nowcasting Hub,
+#' Modified from: https://github.com/KITmetricslab/RESPINOW-Hub/blob/7cce3ae2728116e8c8cc0e4ab29074462c24650e/code/baseline/functions.R#L55 #nolint
+#'
+#'
+#' @param triangle_to_nowcast matrix of the incomplete reporting triangle to be
```
[the remark this answers is not in the document]
Comment (seabbs) [reply R/apply_delay.R:11]: On this I didnt really review for design as I thought we were addressing that in the full design doc (i.e. at a later pass)

File: R/apply_delay.R (hunk 1 above)
Comment (kaitejohnson) [reply R/apply_delay.R]: I modified this to use the `Reduce` function with a separate `calc_expectation()` internal function

File: R/apply_delay.R [hunk 17]
```diff
@@ -0,0 +1,81 @@
...
+apply_delay <- function(triangle_to_nowcast,
+                        delay_pmf) {
+  # Checks that the delay df and the triangle are compatible
+  validate_delay_and_triangle(
+    triangle_to_nowcast,
+    delay_pmf
+  )
+  n_delays <- length(delay_pmf)
+  n_dates <- nrow(triangle_to_nowcast)
+  expectation <- triangle_to_nowcast
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/apply_delay.R:53]: Yep it was, I have removed it.

File: R/apply_delay.R [hunk 18]
```diff
@@ -0,0 +1,81 @@
...
+  return(expectation)
+}
+
+#' Calculate the updates rows of the expected nowcasted triangle
+#'
+#' @param co An integer indicating the column index
+#' @param expectation A matrix of the partially complete reporting triangle
+#' @param n_dates An integer indicating the number of dates in the reporting
+#' triangle (number of rows in the reporting triangle)
+#' @param delay_pmf A vector specifying the probability of a case being
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/apply_delay.R:71]: Good point, good to flag now. I am going to change to no A and just indicating the type capitalised throughout

File: R/apply_delay.R [hunk 19]
```diff
@@ -0,0 +1,81 @@
...
+#'   unname()
+#' delay_df <- get_delay_estimate(
+#'   triangle = triangle_raw,
+#'   max_delay = 20,
+#'   n_history = 30
+#' )
+#' reporting_square <- apply_delay(
+#'   triangle_to_nowcast = triangle_raw,
+#'   delay_pmf = delay_df$pmf
+#' )
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/apply_delay.R:43]: Have changed this throughout.

File: R/apply_delay.R (hunk 9 above)
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/apply_delay.R:32]: Good point, I replaced with a short reporting triangle example.

File: R/validate.R [hunk 20]
```diff
@@ -10,6 +10,13 @@
 validate_triangle <- function(triangle,
                               max_delay,
                               n_history) {
+  # Make sure the input triangle only contains integer values
+  # and is of the correct class
+  checkmate::assert_class(triangle, "matrix")
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/validate.R:15]: Sorry this is so drilled into my head I need to unlearn. Is the ideal usage to use:
`@importFrom checkmate assert_class` in the function documentation, and additionally adding it to the NAMESPACE?

File: tests/testthat/test-apply_delay.R [hunk 21]
```diff
@@ -0,0 +1,126 @@
...
+    # * Note, we will have to do some correction if this is 0, ignore for now*
+    # Finds the expected value for the particular delay by scaling by the
+    # delay pmf for delay d
+    expectation2[(n_dates - co + 2):n_dates, co] <- exp_total * delay_pmf[co]
+  }
+
+  expect_identical(result, expectation2)
+})
+
+test_that("calc_expectation function works correctly", {
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply tests/testthat/test-apply_delay.R:95]: moved this to its own script

File: R/validate.R [hunk 22]
```diff
@@ -47,3 +54,34 @@ validate_triangle <- function(triangle,
...
+#' @description
+#' Various checks to make sure that the reporting triangle  and the delay pmf
+#' passed in to `apply_delay()` are formatted properly and compaitble
+#' @param triangle a matrix of values with rows indicating the time points and
+#' columns indicating the delays
+#' @param delay_pmf a vector of length of the number of delays indicating the
+#' probability of a case being reportined on a given delay
+#'
+#' @returns NULL, invisibly
+validate_delay_and_triangle <- function(triangle,
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/validate.R:68]: Yes will make a separate issue to add tests for the validate functions

File: R/handle_neg_vals.R [hunk 23]
```diff
@@ -12,6 +12,7 @@
 handle_neg_vals <- function(triangle) {
   integer_cols <- seq_len(ncol(triangle))
   pos_triangle <- triangle
+  pos_triangle[is.na(pos_triangle)] <- 0 # Set NAs to 0
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/handle_neg_vals.R:15]: On second thought I think we should do this internally in the function, and then restore it back. I changed this and adjusted the test accordingly.

File: vignettes/baselinenowcast.Rmd (hunk 7 above)
Comment (kaitejohnson) [reply vignettes/baselinenowcast.Rmd:303]: Going to leave for now but can adjust

File: vignettes/baselinenowcast.Rmd (hunk 8 above)
Comment (kaitejohnson) [reply vignettes/baselinenowcast.Rmd:335]: removed

File: R/apply_delay.R [hunk 24]
```diff
@@ -0,0 +1,81 @@
...
+#' nowcasted, with rows representing the time points of reference and columns
+#' representing the delays
+#' @param delay_pmf vector of delays assumed to be indexed starting at the
+#' first delay column in `triangle_to_nowcast`
+#'
+#' @return expectation, a matrix of the same number of rows and columns as the
+#' `triangle_to_nowcast` but with the missing values filled in as point
+#' estimates
+#' @export
+#'
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/apply_delay.R:21]: Making sure I understand, this is because of the space?

File: R/get_delay_estimate.R (hunk 11 above)
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/get_delay_estimate.R:15]: Changed to `.function_name()` for internal functions throughout and using `@keywords internal`

File: tests/testthat/test-apply_delay.R [hunk 25]
```diff
@@ -0,0 +1,126 @@
+test_that("apply_delay function works correctly on simple triangle", {
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply tests/testthat/test-apply_delay.R:1]: No I need to do this, I for sure know these haven't hit all the edge cases we need to think about

File: tests/testthat/test-apply_delay.R [hunk 26]
```diff
@@ -0,0 +1,126 @@
...
+    delay_pmf = delay_pmf
+  )
+
+  expect_is(result, "matrix")
+
+  # Test that the dimensions of the output match the input
+  expect_identical(dim(result), dim(triangle))
+
+  # Test that the known values remain unchanged
+  expect_identical(result[1:3, 1:2], triangle[1:3, 1:2])
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply tests/testthat/test-apply_delay.R:17]: Ok, going to punt this to a separate issue to add additional tests

File: R/apply_delay.R [hunk 27]
```diff
@@ -0,0 +1,81 @@
...
+  n_dates <- nrow(triangle_to_nowcast)
+  expectation <- triangle_to_nowcast
+
+  # Iterates through each column and adds entries to the expected reporting
+  # triangle
+  expectation <- Reduce(function(acc, co) {
+    calc_expectation(co, acc, n_dates, delay_pmf)
+  }, 2:n_delays, init = triangle_to_nowcast)
+
+
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/apply_delay.R:61]: removed

File: vignettes/baselinenowcast.Rmd (hunk 5 above)
Comment (kaitejohnson) [reply vignettes/baselinenowcast.Rmd]: removed

File: R/validate.R (hunk 20 above)
[the remark this answers is not in the document]
Comment (sbfnk) [reply R/validate.R:13]: If you add it in `@importFrom` then `devtools::document()` will add it to NAMESPACE (which you shouldn't have to edit by hand if using roxygen2.

**Review by sbfnk — COMMENTED** [review]
Looks great! At the risk of being overly pernickety the indent in the roxygen docs should be two spaces, see https://style.tidyverse.org/documentation.html#indents-and-line-breaks - for what it's worth if using Rstudio I think you can select the roxygen block and hit Cmd-Shift-/ (with similar shortcuts in emacs and I presume VS code)

**Comment by seabbs** [comment]
Flagging this is in my todos and I can get to it tomorrow.; Happy to defer to @sbfnk if want to merge sooner

File: R/apply_delay.R (hunk 24 above)
[the remark this answers is not in the document]
Comment (seabbs) [reply R/apply_delay.R]: no this is because if  you have a non automated pkgdown reference (which was in the template you started out from) you need to either manually specify functions in it or set up some kind for system to organise them.

File: R/get_delay_estimate.R [hunk 28]
```diff
@@ -1,57 +1,58 @@
...
+#'  estimation. This point estimate of the delay is computed empirically,
+#'  using an iterative algorithm starting from the most recent observations.
+#'  This code was adapted from code written (under an MIT license)
+#'  by the Karlsruhe Institute of Technology RESPINOW
+#'  German Hospitalization Nowcasting Hub.
+#'  Modified from: https://github.com/KITmetricslab/RESPINOW-Hub/blob/7cce3ae2728116e8c8cc0e4ab29074462c24650e/code/baseline/functions.R#L55 #nolint
+#' @param triangle Matrix of the reporting triangle, with rows representing
+#'  the time points of reference and columns representing the delays
+#' @param max_delay Integer indicating the maximum delay to estimate, in units
+#'  of the delay. The default is to use the whole reporting triangle,
```
Comment (seabbs) [inline R/get_delay_estimate.R:15]: Im just here being that guy saying this only has one space and the other docs have two... the only thing that matters is they have the same number to me

File: R/get_delay_estimate.R [hunk 29]
```diff
@@ -1,57 +1,58 @@
...
-#' recent reporting delay. The default is to use the whole reporting triangle,
-#' so `nrow(triangle)-1`
-#' @returns delay_df dataframe of length `max_delay` with columns `delay`
-#' and `pmf`, indicating the point estimate of the empirical probability
-#' mass on each delay
+#' @param n_history Integer indicating the number of reference dates to be
+#'   used in the estimate of the reporting delay, always starting from the most
+#'   recent reporting delay. The default is to use the whole reporting triangle,
+#'   so `nrow(triangle)-1`
+#' @returns delay_df Dataframe of length `max_delay` with columns `delay`
```
Comment (seabbs) [inline R/get_delay_estimate.R:21]: I t
```suggestion
#' @returns delay_df Dataframe with `max_delay` rows and `delay` columns
```

File: R/handle_neg_vals.R [hunk 30]
```diff
@@ -1,17 +1,21 @@
...
-#' as the input triangle, but with negative values of reporting handled via
-#' passing them to the subsequent days delay.
-#' Modified from https://github.com/KITmetricslab/RESPINOW-Hub/blob/main/code/baseline/functions.R #nolint
-#' @param triangle the reporting triangle as a matrix, where rows are the
-#' time points and columns are the delays, already truncated to the maximum
-#' delay and the number of historical observations
-#' @return pos_triangle a positive integer matrix with negative values of
-#' reporting handled via passing them to the subsequent days delay
-handle_neg_vals <- function(triangle) {
+#'   as the input triangle, but with negative values of reporting handled via
```
Comment (seabbs) [inline R/handle_neg_vals.R:4]: here it is three (sorry)

**Review by seabbs — APPROVED** [review]
Content looks good. I see some unresolved points from previous reviews where it is unclear to me what action has been taken. Closing these out will let it auto merge

File: R/get_delay_estimate.R (hunk 28 above)
Comment (kaitejohnson) [reply R/get_delay_estimate.R:15]: fixing

File: R/validate.R [hunk 31]
```diff
@@ -47,3 +54,34 @@ validate_triangle <- function(triangle,
     )
   }
 }
+
+#' Validate triangle to nowcast and delay pmf together
+#' @description
```
[the remark this answers is not in the document]
Comment (kaitejohnson) [reply R/validate.R:57]: I will remove throughout

**Comment by seabbs** [comment]
if you turn on the merge queue and make sure the CI is working it (see EpiNow2) you don't need to keep merging manually
